presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java

```diff
@@ -198,15 +205,32 @@ public void registerBuiltInFunctions(List<? extends BuiltInFunction> functions)
 
     public List<SqlFunction> listFunctions(Session session)
     {
-        Collection<FunctionNamespaceManager<?>> managers = isListBuiltInFunctionsOnly(session) ?
-                ImmutableSet.of(builtInFunctionNamespaceManager) :
-                functionNamespaceManagers.values();
+        Set<String> managerIds = isListBuiltInFunctionsOnly(session) ?
+                ImmutableSet.of(BuiltInFunctionNamespaceManager.ID) :
+                functionNamespaceManagers.keySet();
 
-        return managers.stream()
-                .flatMap(manager -> manager.listFunctions().stream())
-                .filter(function -> function.getVisibility() == PUBLIC ||
-                        (function.getVisibility() == EXPERIMENTAL && SystemSessionProperties.isExperimentalFunctionsEnabled(session)))
-                .collect(toImmutableList());
+        ImmutableList.Builder<SqlFunction> functions = ImmutableList.builder();
+        for (String managerId : managerIds) {
+            FunctionNamespaceManager<? extends SqlFunction> manager = functionNamespaceManagers.get(managerId);
+            Set<CatalogSchemaPrefix> servingNamespaces = this.servingNamespaces.get().get(managerId);
+
+            for (SqlFunction function : manager.listFunctions()) {
+                CatalogSchemaName functionNamespace = function.getSignature().getName().getFunctionNamespace();
+
+                if (!servingNamespaces.contains(functionNamespace.asCatalogSchemaPrefix())
+                        && !servingNamespaces.contains(new CatalogSchemaPrefix(functionNamespace.getCatalogName(), Optional.empty()))) {
+                    throw new PrestoException(
```

Comment:
Hmm, do we need to through error if the function namespace manager has some old functions from a namespace that's no longer used? Personally I prefer the function namespace manager to support a `listFunction(prefix)` API. This way we simply ignore anything that's in the function namespace manager but not configured to be managed by this instance. (Think, the same backend database hosting function namespaces `a.b`, `x.y`, but for one cluster you are only configuring it to manage `a.b`, while on another cluster it's configured to manage `x.y`). I don't see why this needs to be an internal error. It seems to be reasonable. Essentially, the config is constructing a view of the underlying database to be exposed to a certain cluster configuration.